New guide: AWS DynamoDB with Database Access (#19956)
gravitational/teleport: docs/pages/application-access/guides/dynamodb.mdx

---

> title: Access AWS DynamoDB as an Application

Reviewer: Do you think the reader will know what _as an Application_ means?

Author: I don't know myself. I'd welcome any suggestions on how to better differentiate this guide's title from the new one. As discussed in the comment above, the titles should be unique.

Reviewer 2: "Access AWS DynamoDB via App Access"?

Reviewer 2: that may be a bit of redundant phrasing tho

Author: "AWS DynamoDB using App Access"?

Author: See 0fdbfce9c6a73ad77d11b7701b36d4c7ffc73f7d

description: How to access AWS DynamoDB as an Application with Teleport.
---

> Access to AWS DynamoDB can be provided by [**Teleport Application

Reviewer: Didn't you just tell me that `can be` should be avoided in docs? :-)

Author: Yes I did, and I'm happy that it stuck in your brain! In this case I feel that the two cannot be used interchangeably. This is specifically saying that  what's described below is an option, but not the only one.

Access**](../../application-access/introduction.mdx) for the AWS Console and
API. This is an alternative to accessing DynamoDB through the Teleport Database

> service, as described in our [Database Access with AWS DynamoDB](../../database-access/guides/aws-dynamodb.mdx) guide.

Reviewer: Before I read any further, I probably want to understand if I'm on the right guide or if I should click this link and go to the DB access guide.

Maybe a small section describing the pros and cons of each approach (on both pages) would be helpful here?

Author: I agree with you, but lack the subject matter expertise to provide the requested details. Maybe @smallinsky can suggest some points that I can add here?

Reviewer 3: The DynamoDB thought database access mostly works the same but it provides the unified experience with other databases protocols like distinguishable audit logs  and ability to connect through `tsh proxy db`.

Reviewer 2: Main benefit is db access can be used by GUI clients like NoSQL workbench (not possible via app access)

Author: See 0fdbfce9c6a73ad77d11b7701b36d4c7ffc73f7d

---
title: AWS DynamoDB using App Access
description: How to access AWS DynamoDB through the Teleport Application Service
---

Access to AWS DynamoDB can be provided by [**Teleport Application
Access**](../../application-access/introduction.mdx) for the AWS Console and
API. This is an alternative to accessing DynamoDB through the Teleport Database
service, as described in our [Database Access with AWS
DynamoDB](../../database-access/guides/aws-dynamodb.mdx) guide. Note that using

> the Database Access Service will allow you to connect with GUI clients, whereas

Reviewer: DB access also provides richer audit logs. It sounds to me like you should almost always prefer the DB access approach over app access.

> Application Access does not.

Reviewer 2: one advantage is you do not need to configure app service dynamodb for each region you intend to access

> Application Access does not.

Author: Attempted to clarify in 061265850e1867dcc83ad3c2c2106cac84b388f1, @GavinFrazar please review.

> Application Access does not.

Reviewer 2: discussion here: https://github.com/gravitational/teleport/pull/19956/files#r1093692720

marking this resolved so we don't have multiple threads on this

This guide will help you to:

- Install the Teleport Application Service.
- Set up the Teleport Application Service to access the AWS Console and API.
- Connect to your DynamoDB databases through the Teleport Application Service.

<ScopedBlock scope={["oss", "enterprise"]}>
![DynamoDB Self-Hosted](../../../img/application-access/guides/dynamodb_selfhosted.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

> ![DynamoDB Cloud](../../../img/application-access/guides/dynamodb_cloud.png)

Author: Adjusted as described.

---
title: AWS DynamoDB using the Teleport Application Service
description: How to access AWS DynamoDB through the Teleport Application Service
---

Access to AWS DynamoDB can be provided by [**Teleport Application
Access**](../../application-access/introduction.mdx) for the AWS Console and
API. This is an alternative to accessing DynamoDB through the Teleport Database
service, as described in our [Database Access with AWS
DynamoDB](../../database-access/guides/aws-dynamodb.mdx) guide.

> Note that using the Database Service will allow you to connect with GUI clients,
> whereas the Application Service does not. On the other hand, a single Application
> Service node can access DynamoDB across regions, while Database nodes must be
> present in the same region(s) as the DynamoDB database(s).

Reviewer 2: Avoid using "node" as that term is vague. I would rephrase this:

```suggestion
Note that using the Database Service will allow you to connect with GUI clients,
whereas the Application Service does not. On the other hand, a single Teleport application
can be configured to access DynamoDB in any AWS region, while a Teleport Database must be
configured to access DynamoDB in each AWS region.
```

Reviewer 2: And actually I just thought of another one: Database access supports per-session-MFA and hardware backed keys, while application access does not (currently).

Author: I'm ok with not using "node", but the suggested edit still needs some noun, otherwise "application" and "Database" becomes the noun. Since, in the latter case for example, it's not an actual database that teleport runs. How about:

```suggestion
Note that using the Database Service will allow you to connect with GUI clients,
whereas the Application Service does not. On the other hand, a single Application
Service instance can access DynamoDB across regions, while Teleport Database
instances must be present in the same region(s) as the DynamoDB database(s).
```

Reviewer 2: how about "resource" instead of "instance"? I just don't the wording here to sound like you need an agent running in the same AWS region as the dynamodb endpoint.

Author: OK, I think I misunderstood [this earlier comment](https://github.com/gravitational/teleport/pull/19956#discussion_r1086015392):

> one advantage is you do not need to configure app service dynamodb for each region you intend to access

I _thought_ this implied that database access did require a service in each region. Can we clarify what this means? I tagged you for review of that change in that thread as well.

Reviewer 2: "service" refers to the agent in my opinion. It's the `db_service` section of the yaml.

You do not need multiple agents (services). You need one configured database per region. They can all be configured on one agent.

Author: Clarified in https://github.com/gravitational/teleport/pull/19956/commits/f64a9fa3a51b5bbee592d177afe76a9e1799060a
